## Data Breach Analytics 2005 - 2017

**By Miriam Rodriguez**


### Data Description

#### Dataset
- Dataset name:              Privacy_Rights_Clearinghouse-Data-Breaches-Export.csv
- Source:                    https://www.privacyrights.org/data-breaches (Links to an external site.). 
- Data Description and FAQ:  https://www.privacyrights.org/chronology-data-breaches-faq 
- Duration of the data:      2005 through 2017

#### Data Breach Types
   - CARD - Payment Card Fraud:    Fraud involving debit and credit cards that is not accomplished via hacking (e.g. skimming                                      devices at point-of-service terminals).
   - HACK - Hacking or Malware:    Hacked by outside party or infected by malware
   - INSD - Insider:               Someone with legitimate access intentionally breaches information, such as an employee,                                          contractor, or customer)
   - PHYS - Physical Loss:         Includes paper documents that are lost, discarded, or stolen (non-electronic)
   - PORT - Portable Device:       Lost, discarded, or stolen laptop, PDA, smartphone, memory stick, CDs, hard drive, data tape,                                    etc.
   - STAT - Stationary Device:     Stationary computer loss (lost, inappropriately accessed, discarded, or stolen computer or                                      server not designed for mobility)
   - DISC - Unintended Disclosure: Unintended disclosure (not involving hacking, intentional breach, or physical loss i.e.                                          sensitive information posted publicly, mishandled, sent to the wrong party via publishing                                      online, sending in an email, sending in a mailing or sending via fax.
   - Unknown

#### Institution/Organization) Type
-	BSF - Businesses - Financial and Insurance Services
-	BSO - Businesses - Other
-	BSR - Businesses - Retail/Merchant – Including Online Retail
-	EDU - Educational Institutions
-	GOV - Government & Military
-	MED - Healthcare - Medical Providers & Medical Insurance Services
-	NGO - Nonprofit Organizations

#### Data elements/column names
1.	Date Made Public: Date Breach information released to public (date: year, month, day)
2.	Company: Company breached (text)
3.	City: City of breached company (text)
4.	State: State of breached company (text)
5.	Type of Breach: Refer to four-character Breach Type above
6.	Type of Organization: Refer to three-character Institution/Organization Type above
7.	Total Records: Number of records breached (integer)
8.	Description of Incident: Text describing breach (text)
9.	Information Source: Location of database source (text)
10.	Source URL: Location of data source URL (text)
11.	Year of Breach: Four digit year (numeric)
12.	Latitude: Location Latitude (signed numeric long)
13.	Longitude: Location Longitude (signed numeric long)


In [1]:
# import python packages
import pandas as pd
import numpy as np
import matplotlib as plt
%matplotlib inline

# Importing Data & Processing (ETL)

In [2]:
#import breach data ... open or read the breach data
df=pd.read_csv("Privacy_Rights_Clearinghouse-Data-Breaches-Export.csv")
df.head()


Date Made Public                              Company             City  \
0        16-May-08  Greil Memorial Psychiatric Hospital       Montgomery   
1        21-Mar-08                         Compass Bank       Birmingham   
2         7-Aug-07              Electronic Data Systems       Montgomery   
3         3-Jun-07      Gadsden State Community College  College Gadsden   
4         5-Apr-07                   DCH Health Systems       Tuscaloosa   

     State Type of breach Type of organization Total Records  \
0  Alabama           PHYS                  EDU             0   
1  Alabama           INSD                  BSF     1,000,000   
2  Alabama           INSD                  BSO           498   
3  Alabama           PHYS                  EDU           400   
4  Alabama           PORT                  MED         6,000   

                             Description of incident Information Source  \
0  Index cards containing patients \n            ...        Dataloss DB   
1  A database containing names, account \n       ...        Dataloss DB   
2  A former employee \n            was arrested t...        Dataloss DB   
3  Students who took \n            an Art Appreci...        Dataloss DB   
4  An encrypted disc \n            and hardcopy d...        Dataloss DB   

  Source URL  Year of Breach   Latitude  Longitude  
0        NaN          2008.0  32.366805 -86.299969  
1        NaN          2008.0  33.520661 -86.802490  
2        NaN          2007.0  32.366805 -86.299969  
3        NaN          2007.0  34.025272 -85.995891  
4        NaN          2007.0  33.209841 -87.569174

### Determine data types and missing values

In [3]:
# how many missing values in each column or variable
df.isnull().sum()

Date Made Public              0
Company                       0
City                       2520
State                        68
Type of breach                0
Type of organization          0
Total Records                38
Description of incident       3
Information Source           54
Source URL                 5410
Year of Breach               33
Latitude                      0
Longitude                     0
dtype: int64

In [4]:
# Describe data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8202 entries, 0 to 8201
Data columns (total 13 columns):
Date Made Public           8202 non-null object
Company                    8202 non-null object
City                       5682 non-null object
State                      8134 non-null object
Type of breach             8202 non-null object
Type of organization       8202 non-null object
Total Records              8164 non-null object
Description of incident    8199 non-null object
Information Source         8148 non-null object
Source URL                 2792 non-null object
Year of Breach             8169 non-null float64
Latitude                   8202 non-null float64
Longitude                  8202 non-null float64
dtypes: float64(3), object(10)
memory usage: 833.1+ KB


**There are 8202 records total.**  

** There are data quality issues**


** Based upon above counts, there are missing values for City, State, Total_Recs, Description, Breach_Year **


Column name: City 
- How resolved: Drop column. 
- Justification: The majority of cities are not provided.  All we really need is State.

Column name: State
- How resolved: Replace spaces with 'United States'
- Justification: Since this value was not provided in the file, the assumption is that the breach was national.

Column name: Total_Recs 
- How resolved: Move zeros to null values. Then convert to float to remove zeros, then convert to int.
- Justification: This will enable the ability to measure the quanitative impact.

Column name: Description 
- How resolved: Replace spaces with 'None'
- Justification: This value was not provided in the file.

Column name: Breach_Year
- How resolved:   Drop column
- Justification:  We have date with year with no empty values so do not need this field


Remove the following fields as they are not needed:
- Information Source 
- Source URL 

In [5]:
#Two columns (Information Source, Source URL) are not necessary for analysis. They will be dropped.
df = df.drop(['Information Source', 'Source URL'], axis=1)
df.head()

Date Made Public                              Company             City  \
0        16-May-08  Greil Memorial Psychiatric Hospital       Montgomery   
1        21-Mar-08                         Compass Bank       Birmingham   
2         7-Aug-07              Electronic Data Systems       Montgomery   
3         3-Jun-07      Gadsden State Community College  College Gadsden   
4         5-Apr-07                   DCH Health Systems       Tuscaloosa   

     State Type of breach Type of organization Total Records  \
0  Alabama           PHYS                  EDU             0   
1  Alabama           INSD                  BSF     1,000,000   
2  Alabama           INSD                  BSO           498   
3  Alabama           PHYS                  EDU           400   
4  Alabama           PORT                  MED         6,000   

                             Description of incident  Year of Breach  \
0  Index cards containing patients \n            ...          2008.0   
1  A database containing names, account \n       ...          2008.0   
2  A former employee \n            was arrested t...          2007.0   
3  Students who took \n            an Art Appreci...          2007.0   
4  An encrypted disc \n            and hardcopy d...          2007.0   

    Latitude  Longitude  
0  32.366805 -86.299969  
1  33.520661 -86.802490  
2  32.366805 -86.299969  
3  34.025272 -85.995891  
4  33.209841 -87.569174

Remove spaces from field names (rename)
- Total Records:           Total_Recs
- Description of incident: Description
- Type of organization:    Organization_Type
- Type of breach:          Breach_Type
- Date Made Public         Date_Public

In [6]:
#Drop column Breach_Year which has missing values. No longer needed since Date_Public has no missing values.
df = df.drop(['Year of Breach'], axis=1)
df.head()

Date Made Public                              Company             City  \
0        16-May-08  Greil Memorial Psychiatric Hospital       Montgomery   
1        21-Mar-08                         Compass Bank       Birmingham   
2         7-Aug-07              Electronic Data Systems       Montgomery   
3         3-Jun-07      Gadsden State Community College  College Gadsden   
4         5-Apr-07                   DCH Health Systems       Tuscaloosa   

     State Type of breach Type of organization Total Records  \
0  Alabama           PHYS                  EDU             0   
1  Alabama           INSD                  BSF     1,000,000   
2  Alabama           INSD                  BSO           498   
3  Alabama           PHYS                  EDU           400   
4  Alabama           PORT                  MED         6,000   

                             Description of incident   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [7]:
# rename columns
df = df.rename(columns={'Total Records': 'Total_Recs'})
df.head()

Date Made Public                              Company             City  \
0        16-May-08  Greil Memorial Psychiatric Hospital       Montgomery   
1        21-Mar-08                         Compass Bank       Birmingham   
2         7-Aug-07              Electronic Data Systems       Montgomery   
3         3-Jun-07      Gadsden State Community College  College Gadsden   
4         5-Apr-07                   DCH Health Systems       Tuscaloosa   

     State Type of breach Type of organization Total_Recs  \
0  Alabama           PHYS                  EDU          0   
1  Alabama           INSD                  BSF  1,000,000   
2  Alabama           INSD                  BSO        498   
3  Alabama           PHYS                  EDU        400   
4  Alabama           PORT                  MED      6,000   

                             Description of incident   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [8]:
# rename columns Description of incident: Description
df = df.rename(columns={'Description of incident': 'Description'})

In [9]:
# rename columns Type of organization:    Organization_Type
df = df.rename(columns={'Type of organization': 'Organization_Type'})

In [10]:
# rename columns Date Made Public:    Date_Public
df = df.rename(columns={'Date Made Public': 'Date_Public'})

In [11]:
# rename columns Type of breach:          Breach_Type
df = df.rename(columns={'Type of breach': 'Breach_Type'})
df.head()

Date_Public                              Company             City    State  \
0   16-May-08  Greil Memorial Psychiatric Hospital       Montgomery  Alabama   
1   21-Mar-08                         Compass Bank       Birmingham  Alabama   
2    7-Aug-07              Electronic Data Systems       Montgomery  Alabama   
3    3-Jun-07      Gadsden State Community College  College Gadsden  Alabama   
4    5-Apr-07                   DCH Health Systems       Tuscaloosa  Alabama   

  Breach_Type Organization_Type Total_Recs  \
0        PHYS               EDU          0   
1        INSD               BSF  1,000,000   
2        INSD               BSO        498   
3        PHYS               EDU        400   
4        PORT               MED      6,000   

                                         Description   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

*First, need to resolve missing Year of Breach (since date made public all there, can convert to date and use that year).  Convert the date from an object to a date timestamp and only use year.*

In [12]:
df['Date_Public'] =  pd.to_datetime(df['Date_Public'], errors='coerce').dt.year

In [13]:
# rename columns Date_Public:    Breach_Year
df = df.rename(columns={'Date_Public': 'Breach_Year'})

In [14]:
df.head()

Breach_Year                              Company             City    State  \
0         2008  Greil Memorial Psychiatric Hospital       Montgomery  Alabama   
1         2008                         Compass Bank       Birmingham  Alabama   
2         2007              Electronic Data Systems       Montgomery  Alabama   
3         2007      Gadsden State Community College  College Gadsden  Alabama   
4         2007                   DCH Health Systems       Tuscaloosa  Alabama   

  Breach_Type Organization_Type Total_Recs  \
0        PHYS               EDU          0   
1        INSD               BSF  1,000,000   
2        INSD               BSO        498   
3        PHYS               EDU        400   
4        PORT               MED      6,000   

                                         Description   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [15]:
# replace null value with zero 
df = df.fillna({'Total_Recs': '0'})
df.isnull().sum()

Breach_Year             0
Company                 0
City                 2520
State                  68
Breach_Type             0
Organization_Type       0
Total_Recs              0
Description             3
Latitude                0
Longitude               0
dtype: int64

In [16]:
#Convert Total_Recs to float and remove zeros
df["Total_Recs"] = df["Total_Recs"].str.replace(",","").astype(float)

In [17]:
#Convert Total_Recs to integer
df['Total_Recs'] = df['Total_Recs'].astype(np.int64)

In [18]:
df.head()

Breach_Year                              Company             City    State  \
0         2008  Greil Memorial Psychiatric Hospital       Montgomery  Alabama   
1         2008                         Compass Bank       Birmingham  Alabama   
2         2007              Electronic Data Systems       Montgomery  Alabama   
3         2007      Gadsden State Community College  College Gadsden  Alabama   
4         2007                   DCH Health Systems       Tuscaloosa  Alabama   

  Breach_Type Organization_Type  Total_Recs  \
0        PHYS               EDU           0   
1        INSD               BSF     1000000   
2        INSD               BSO         498   
3        PHYS               EDU         400   
4        PORT               MED        6000   

                                         Description   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [19]:
#Drop column City which has missing values. Not needed as we are matching GDP by State and the majority of the cities are missing
df = df.drop(['City'], axis=1)
df.head()

Breach_Year                              Company    State Breach_Type  \
0         2008  Greil Memorial Psychiatric Hospital  Alabama        PHYS   
1         2008                         Compass Bank  Alabama        INSD   
2         2007              Electronic Data Systems  Alabama        INSD   
3         2007      Gadsden State Community College  Alabama        PHYS   
4         2007                   DCH Health Systems  Alabama        PORT   

  Organization_Type  Total_Recs  \
0               EDU           0   
1               BSF     1000000   
2               BSO         498   
3               EDU         400   
4               MED        6000   

                                         Description   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [20]:
# replace null value with 'United States' 
df = df.fillna({'State': 'United States'})

In [21]:
df.groupby('State').size()

State
Alabama                   78
Alaska                    23
Arizona                  135
Arkansas                  54
Beijing                    1
Berlin                     1
British Columbia           3
Buckinghamshire            2
California              1287
Cheshire                   1
Colorado                 165
Connecticut              138
Delaware                  22
District Of Columbia     152
Dublin                     1
Florida                  435
Georgia                  244
Grand Bahama               1
Guangdong                  1
Hawaii                    27
Idaho                     22
Illinois                 335
Indiana                  208
Iowa                      67
Kansas                    54
Kentucky                 112
London                     2
Louisiana                 61
Maine                     33
Maryland                 374
                        ... 
Nebraska                  38
Nevada                    62
New Hampshire             44
New Jers

##### There are 'states' that are not part of the US.  These need to either be set to 'Non US' or deleted from this study as this are US Breaches only.

In [22]:
df = df[df.State != 'Beijing']
df = df[df.State != 'Berlin']
df = df[df.State != 'British Columbia']
df = df[df.State != 'Buckinghamshire']
df = df[df.State != 'Cheshire']
df = df[df.State != 'Dublin']
df = df[df.State != 'Grand Bahama']
df = df[df.State != 'Guangdong']
df = df[df.State != 'London']
df = df[df.State != 'Noord Holland']
df = df[df.State != 'Quebec']
df = df[df.State != 'Tokyo']
df = df[df.State != 'Ontario']


In [23]:
## Recheck states
df.groupby('State').size()

State
Alabama                   78
Alaska                    23
Arizona                  135
Arkansas                  54
California              1287
Colorado                 165
Connecticut              138
Delaware                  22
District Of Columbia     152
Florida                  435
Georgia                  244
Hawaii                    27
Idaho                     22
Illinois                 335
Indiana                  208
Iowa                      67
Kansas                    54
Kentucky                 112
Louisiana                 61
Maine                     33
Maryland                 374
Massachusetts            233
Michigan                 147
Minnesota                144
Mississippi               35
Missouri                 137
Montana                   34
Nebraska                  38
Nevada                    62
New Hampshire             44
New Jersey               154
New Mexico                50
New York                 602
North Carolina           204
North Da

In [24]:
# replace null value with 'not specified' 
df = df.fillna({'Description': 'None'})
df.isnull().sum()

Breach_Year          0
Company              0
State                0
Breach_Type          0
Organization_Type    0
Total_Recs           0
Description          0
Latitude             0
Longitude            0
dtype: int64

In [25]:
df.head()

Breach_Year                              Company    State Breach_Type  \
0         2008  Greil Memorial Psychiatric Hospital  Alabama        PHYS   
1         2008                         Compass Bank  Alabama        INSD   
2         2007              Electronic Data Systems  Alabama        INSD   
3         2007      Gadsden State Community College  Alabama        PHYS   
4         2007                   DCH Health Systems  Alabama        PORT   

  Organization_Type  Total_Recs  \
0               EDU           0   
1               BSF     1000000   
2               BSO         498   
3               EDU         400   
4               MED        6000   

                                         Description   Latitude  Longitude  
0  Index cards containing patients \n            ...  32.366805 -86.299969  
1  A database containing names, account \n       ...  33.520661 -86.802490  
2  A former employee \n            was arrested t...  32.366805 -86.299969  
3  Students who took \n            an Art Appreci...  34.025272 -85.995891  
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174

In [26]:
# Create column with numerical values for Breach_Type
# changing categorical values with map
df['Breach_Type_CAT'] = df['Breach_Type'].map({'CARD': '1', 'HACK': '2', 'INSD': '3', 'PHYS': '4', 'PORT': '5', 'STAT': '6', 'DISC': '7'})
df.head()

Breach_Year                              Company    State Breach_Type  \
0         2008  Greil Memorial Psychiatric Hospital  Alabama        PHYS   
1         2008                         Compass Bank  Alabama        INSD   
2         2007              Electronic Data Systems  Alabama        INSD   
3         2007      Gadsden State Community College  Alabama        PHYS   
4         2007                   DCH Health Systems  Alabama        PORT   

  Organization_Type  Total_Recs  \
0               EDU           0   
1               BSF     1000000   
2               BSO         498   
3               EDU         400   
4               MED        6000   

                                         Description   Latitude  Longitude  \
0  Index cards containing patients \n            ...  32.366805 -86.299969   
1  A database containing names, account \n       ...  33.520661 -86.802490   
2  A former employee \n            was arrested t...  32.366805 -86.299969   
3  Students who took \n            an Art Appreci...  34.025272 -85.995891   
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174   

  Breach_Type_CAT  
0               4  
1               3  
2               3  
3               4  
4               5

In [27]:
# Create column with numerical values for Organization_Type
# changing categorical values with map
df['Organization_Type_CAT'] = df['Organization_Type'].map({'BSF': '1', 'BSO': '2', 'BSR': '3', 'EDU': '4', 'GOV': '5', 'MED': '6', 'NGO': '7'})
df.head()

Breach_Year                              Company    State Breach_Type  \
0         2008  Greil Memorial Psychiatric Hospital  Alabama        PHYS   
1         2008                         Compass Bank  Alabama        INSD   
2         2007              Electronic Data Systems  Alabama        INSD   
3         2007      Gadsden State Community College  Alabama        PHYS   
4         2007                   DCH Health Systems  Alabama        PORT   

  Organization_Type  Total_Recs  \
0               EDU           0   
1               BSF     1000000   
2               BSO         498   
3               EDU         400   
4               MED        6000   

                                         Description   Latitude  Longitude  \
0  Index cards containing patients \n            ...  32.366805 -86.299969   
1  A database containing names, account \n       ...  33.520661 -86.802490   
2  A former employee \n            was arrested t...  32.366805 -86.299969   
3  Students who took \n            an Art Appreci...  34.025272 -85.995891   
4  An encrypted disc \n            and hardcopy d...  33.209841 -87.569174   

  Breach_Type_CAT Organization_Type_CAT  
0               4                     4  
1               3                     1  
2               3                     2  
3               4                     4  
4               5                     6

In [28]:
# replace null value with zero 
df = df.fillna({'Breach_Type_CAT': '0'})
df.isnull().sum()

Breach_Year              0
Company                  0
State                    0
Breach_Type              0
Organization_Type        0
Total_Recs               0
Description              0
Latitude                 0
Longitude                0
Breach_Type_CAT          0
Organization_Type_CAT    0
dtype: int64

In [29]:
#Convert Category to integer
df['Breach_Type_CAT'] = df['Breach_Type_CAT'].astype(np.int64)

In [30]:
#Convert Category to integer
df['Organization_Type_CAT'] = df['Organization_Type_CAT'].astype(np.int64)

In [31]:
#show the information about the dataset - no missing data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8177 entries, 0 to 8201
Data columns (total 11 columns):
Breach_Year              8177 non-null int64
Company                  8177 non-null object
State                    8177 non-null object
Breach_Type              8177 non-null object
Organization_Type        8177 non-null object
Total_Recs               8177 non-null int64
Description              8177 non-null object
Latitude                 8177 non-null float64
Longitude                8177 non-null float64
Breach_Type_CAT          8177 non-null int64
Organization_Type_CAT    8177 non-null int64
dtypes: float64(2), int64(4), object(5)
memory usage: 766.6+ KB


# Data Understanding 

## Breach Statistics and ETL

**Business questions include comparisons of data breaches in terms of U.S. states.**  
- Larger states (e.g., California) with many businesses/populations have more cases of data breaches than U.S states such as       Wyoming and Idaho. 
- California' is shows that it is the top state for breaches reported, normalizing the frequency by each state's GDP will         provide a more accurate statistic for States. 
- Data about gross domestic product (GDP) for U.S. states between 2005 and 2017 was obtained from a .csv download.  
- Create a new dataframe with data breaches by state by combining state and year data from the breach file and the GDP file. 
- Normalize the number of data breaches by using each state's GDP.  

In [32]:
df.groupby(['State']).size()

State
Alabama                   78
Alaska                    23
Arizona                  135
Arkansas                  54
California              1287
Colorado                 165
Connecticut              138
Delaware                  22
District Of Columbia     152
Florida                  435
Georgia                  244
Hawaii                    27
Idaho                     22
Illinois                 335
Indiana                  208
Iowa                      67
Kansas                    54
Kentucky                 112
Louisiana                 61
Maine                     33
Maryland                 374
Massachusetts            233
Michigan                 147
Minnesota                144
Mississippi               35
Missouri                 137
Montana                   34
Nebraska                  38
Nevada                    62
New Hampshire             44
New Jersey               154
New Mexico                50
New York                 602
North Carolina           204
North Da

## Create numerical representations of states and years for statistical purposes

In [33]:
df['State_CAT'] = df['State'].map({'Wyoming' : '52', 'Wisconsin' : '51', 'West Virginia' : '50', 'Washington' : '49', 'Virginia' : '48', 'Vermont' : '47', 'Utah' : '46', 'Texas' : '45', 'Tennessee' : '44', 'South Dakota' : '43', 'South Carolina' : '42', 'Rhode Island' : '41', 'Puerto Rico' : '40', 'Pennsylvania' : '39', 'Oregon' : '38', 'Oklahoma' : '37', 'Ohio' : '36', 'North Dakota' : '35', 'North Carolina' : '34', 'New York' : '33', 'New Mexico' : '32', 'New Jersey' : '31', 'New Hampshire' : '30', 'Nevada' : '29', 'Nebraska' : '28', 'Montana' : '27', 'Missouri' : '26', 'Mississippi' : '25', 'Minnesota' : '24', 'Michigan' : '23', 'Massachusetts' : '22', 'Maryland' : '21', 'Maine' : '20', 'Louisiana' : '19', 'Kentucky' : '18', 'Kansas' : '17', 'Iowa' : '16', 'Indiana' : '15', 'Illinois' : '14', 'Idaho' : '13', 'Hawaii' : '12', 'Georgia' : '11', 'Florida' : '10', 'District Of Columbia' : '9', 'Delaware': '8', 'United States' : '0', 'Alabama': '1', 'Alaska': '2', 'Arizona': '3', 'Arkansas': '4', 'California': '5', 'Colorado': '6', 'Connecticut': '7'})

In [34]:
# Convert year to string to get meaningful statistics
df['Breach_Year'] = df['Breach_Year'].astype(str)

### Add numeric category for Breach_Year

In [35]:
df['Year_CAT'] = df['Breach_Year'].map({'2005' : '1', '2006' : '2', '2007' : '3', '2008' : '4', '2009' : '5', '2010' : '6', '2011' : '7', '2012' : '8', '2013' : '9', '2014' : '10', '2015' : '11', '2016' : '12', '2017' : '13', '2018' : '14'})

In [36]:
#Convert Category to integer
df['State_CAT'] = df['State_CAT'].astype(np.int64)

In [37]:
#Convert Category to integer
df['Year_CAT'] = df['Year_CAT'].astype(np.int64)

In [38]:
df = df.sort_values(["State_CAT","Year_CAT"], ascending=True)  
df.head()

Breach_Year                                         Company  \
8136        2006                       Deloitte & Touche, McAfee   
8134        2007                   TennCare, Americhoice Inc.      
8139        2007                      Mercer Health and Benefits   
8140        2007  Fidelity Investments, Dairy Farmers of America   
8141        2007                                      Dai Nippon   

              State Breach_Type Organization_Type  Total_Recs  \
8136  United States        PORT               BSO        9290   
8134  United States        PORT               MED       67000   
8139  United States        PORT               BSF       10500   
8140  United States        PORT               BSF          69   
8141  United States        INSD               BSO           0   

                                            Description   Latitude  Longitude  \
8136  An external auditor lost a CD with names, Soci...  37.090240 -95.712891   
8134  There are 67,000 TennCare \n            enroll...  35.960638 -83.920739   
8139     A laptop computer was stolen from a vehicle.    37.090240 -95.712891   
8140  A Fidelity laptop used by a former Fidelity em...  37.090240 -95.712891   
8141  A former contract worker of a Japanese commerc...  37.090240 -95.712891   

      Breach_Type_CAT  Organization_Type_CAT  State_CAT  Year_CAT  
8136                5                      2          0         2  
8134                5                      6          0         3  
8139                5                      1          0         3  
8140                5                      1          0         3  
8141                3                      2          0         3

In [39]:
df['Breach_Occurrence'] = 1

In [40]:
df.head()

Breach_Year                                         Company  \
8136        2006                       Deloitte & Touche, McAfee   
8134        2007                   TennCare, Americhoice Inc.      
8139        2007                      Mercer Health and Benefits   
8140        2007  Fidelity Investments, Dairy Farmers of America   
8141        2007                                      Dai Nippon   

              State Breach_Type Organization_Type  Total_Recs  \
8136  United States        PORT               BSO        9290   
8134  United States        PORT               MED       67000   
8139  United States        PORT               BSF       10500   
8140  United States        PORT               BSF          69   
8141  United States        INSD               BSO           0   

                                            Description   Latitude  Longitude  \
8136  An external auditor lost a CD with names, Soci...  37.090240 -95.712891   
8134  There are 67,000 TennCare \n            enroll...  35.960638 -83.920739   
8139     A laptop computer was stolen from a vehicle.    37.090240 -95.712891   
8140  A Fidelity laptop used by a former Fidelity em...  37.090240 -95.712891   
8141  A former contract worker of a Japanese commerc...  37.090240 -95.712891   

      Breach_Type_CAT  Organization_Type_CAT  State_CAT  Year_CAT  \
8136                5                      2          0         2   
8134                5                      6          0         3   
8139                5                      1          0         3   
8140                5                      1          0         3   
8141                3                      2          0         3   

      Breach_Occurrence  
8136                  1  
8134                  1  
8139                  1  
8140                  1  
8141                  1

In [41]:
# Find out how many state values there are, looks like 53.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8177 entries, 8136 to 8133
Data columns (total 14 columns):
Breach_Year              8177 non-null object
Company                  8177 non-null object
State                    8177 non-null object
Breach_Type              8177 non-null object
Organization_Type        8177 non-null object
Total_Recs               8177 non-null int64
Description              8177 non-null object
Latitude                 8177 non-null float64
Longitude                8177 non-null float64
Breach_Type_CAT          8177 non-null int64
Organization_Type_CAT    8177 non-null int64
State_CAT                8177 non-null int64
Year_CAT                 8177 non-null int64
Breach_Occurrence        8177 non-null int64
dtypes: float64(2), int64(6), object(6)
memory usage: 958.2+ KB


## GDP file transformation-  Import gdp data ... open or read the gdp data

### Source of data : https://www.bea.gov/regional/downloadzip.cfm

In [42]:
gp=pd.read_csv("gdpstate_naics_all/gdpstate_naics_all.csv")
gp.head()


GeoFIPS        GeoName Region ComponentId  \
0       0  United States    NaN         200   
1       0  United States    NaN         200   
2       0  United States    NaN         200   
3       0  United States    NaN         200   
4       0  United States    NaN         200   

                                       ComponentName IndustryId  \
0  Gross domestic product (GDP) by state (million...          1   
1  Gross domestic product (GDP) by state (million...          2   
2  Gross domestic product (GDP) by state (million...          3   
3  Gross domestic product (GDP) by state (million...          4   
4  Gross domestic product (GDP) by state (million...          5   

  IndustryClassification                                    Description  \
0                    ...                             All industry total   
1                    ...                             Private industries   
2                     11    Agriculture, forestry, fishing, and hunting   
3                111-112                                          Farms   
4                113-115      Forestry, fishing, and related activities   

      1997     1998    ...         2008      2009      2010      2011  \
0  8542530  9024434    ...     14626598  14320114  14859772  15406002   
1  7459395  7894015    ...     12716179  12352979  12826507  13348439   
2   108796    99940    ...       154525    137655    160217    197241   
3    88136    79030    ...       126345    109800    129725    166249   
4    20660    20910    ...        28180     27855     30492     30992   

       2012      2013      2014      2015      2016      2017  
0  16041243  16576738  17312308  18007206  18509998  19263350  
1  13957545  14468465  15149621  15776274  16224645  16925936  
2    185800    221821    204404    184791    177580    173445  
3    151489    186960    167709    145476    136672      (NA)  
4     34311     34861     36695     39315     40907      (NA)  

[5 rows x 29 columns]

#### Keeping GeoName, Componentid (to select by Componentid later) and years 2005 - 2017.  
#### All other columns will be dropped.  
#### Only Componentid of 200 will be used. 
#### Will group and sum by GeoName, then match with data by year.

In [43]:
gp = gp.drop(['GeoFIPS', 'Region', 'ComponentName', 'IndustryId', 'IndustryClassification', 'Description'], axis=1)
gp.head()

GeoName ComponentId     1997     1998     1999      2000      2001  \
0  United States         200  8542530  9024434  9595951  10219801  10556006   
1  United States         200  7459395  7894015  8403164   8956010   9223548   
2  United States         200   108796    99940    92808     98517    100027   
3  United States         200    88136    79030    70934     76044     78094   
4  United States         200    20660    20910    21874     22473     21933   

       2002      2003      2004    ...         2008      2009      2010  \
0  10910833  11440109  12200160    ...     14626598  14320114  14859772   
1   9499527   9950967  10627989    ...     12716179  12352979  12826507   
2     95800    116075    142695    ...       154525    137655    160217   
3     74033     93102    118908    ...       126345    109800    129725   
4     21767     22973     23787    ...        28180     27855     30492   

       2011      2012      2013      2014      2015      2016      2017  
0  15406002  16041243  16576738  17312308  18007206  18509998  19263350  
1  13348439  13957545  14468465  15149621  15776274  16224645  16925936  
2    197241    185800    221821    204404    184791    177580    173445  
3    166249    151489    186960    167709    145476    136672      (NA)  
4     30992     34311     34861     36695     39315     40907      (NA)  

[5 rows x 23 columns]

In [44]:
gp = gp.drop(['1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004'], axis=1)
gp.head()

GeoName ComponentId      2005      2006      2007      2008  \
0  United States         200  13015161  13773226  14391149  14626598   
1  United States         200  11366297  12045446  12572387  12716179   
2  United States         200    128571    128345    141999    154525   
3  United States         200    104514     99352    113533    126345   
4  United States         200     24057     28993     28466     28180   

       2009      2010      2011      2012      2013      2014      2015  \
0  14320114  14859772  15406002  16041243  16576738  17312308  18007206   
1  12352979  12826507  13348439  13957545  14468465  15149621  15776274   
2    137655    160217    197241    185800    221821    204404    184791   
3    109800    129725    166249    151489    186960    167709    145476   
4     27855     30492     30992     34311     34861     36695     39315   

       2016      2017  
0  18509998  19263350  
1  16224645  16925936  
2    177580    173445  
3    136672      (NA)  
4     40907      (NA)

In [45]:
# Filter rows using a criterion (GDP only ) = 200

new_gp = gp[gp.ComponentId == '200']
new_gp.head()

GeoName ComponentId      2005      2006      2007      2008  \
0  United States         200  13015161  13773226  14391149  14626598   
1  United States         200  11366297  12045446  12572387  12716179   
2  United States         200    128571    128345    141999    154525   
3  United States         200    104514     99352    113533    126345   
4  United States         200     24057     28993     28466     28180   

       2009      2010      2011      2012      2013      2014      2015  \
0  14320114  14859772  15406002  16041243  16576738  17312308  18007206   
1  12352979  12826507  13348439  13957545  14468465  15149621  15776274   
2    137655    160217    197241    185800    221821    204404    184791   
3    109800    129725    166249    151489    186960    167709    145476   
4     27855     30492     30992     34311     34861     36695     39315   

       2016      2017  
0  18509998  19263350  
1  16224645  16925936  
2    177580    173445  
3    136672      (NA)  
4     40907      (NA)

In [46]:
new_gp = new_gp.drop(['ComponentId'], axis=1)
new_gp.head()

GeoName      2005      2006      2007      2008      2009      2010  \
0  United States  13015161  13773226  14391149  14626598  14320114  14859772   
1  United States  11366297  12045446  12572387  12716179  12352979  12826507   
2  United States    128571    128345    141999    154525    137655    160217   
3  United States    104514     99352    113533    126345    109800    129725   
4  United States     24057     28993     28466     28180     27855     30492   

       2011      2012      2013      2014      2015      2016      2017  
0  15406002  16041243  16576738  17312308  18007206  18509998  19263350  
1  13348439  13957545  14468465  15149621  15776274  16224645  16925936  
2    197241    185800    221821    204404    184791    177580    173445  
3    166249    151489    186960    167709    145476    136672      (NA)  
4     30992     34311     34861     36695     39315     40907      (NA)

In [47]:
# rename columns Type of organization:    Organization_Type
new_gp = new_gp.rename(columns={'GeoName': 'State'})

In [48]:
#check to see if values are numeric
new_gp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5400 entries, 0 to 5399
Data columns (total 14 columns):
State    5400 non-null object
2005     5400 non-null object
2006     5400 non-null object
2007     5400 non-null object
2008     5400 non-null object
2009     5400 non-null object
2010     5400 non-null object
2011     5400 non-null object
2012     5400 non-null object
2013     5400 non-null object
2014     5400 non-null object
2015     5400 non-null object
2016     5400 non-null object
2017     5400 non-null object
dtypes: object(14)
memory usage: 632.8+ KB


In [49]:
new_gp['2017'] = new_gp['2017'].replace(['(NA)'], '0')
new_gp['2016'] = new_gp['2016'].replace(['(NA)'], '0')
new_gp['2015'] = new_gp['2015'].replace(['(NA)'], '0')
new_gp['2014'] = new_gp['2014'].replace(['(NA)'], '0')
new_gp['2013'] = new_gp['2013'].replace(['(NA)'], '0')
new_gp['2012'] = new_gp['2012'].replace(['(NA)'], '0')
new_gp['2011'] = new_gp['2011'].replace(['(NA)'], '0')
new_gp['2010'] = new_gp['2010'].replace(['(NA)'], '0')
new_gp['2009'] = new_gp['2009'].replace(['(NA)'], '0')
new_gp['2008'] = new_gp['2008'].replace(['(NA)'], '0')
new_gp['2007'] = new_gp['2007'].replace(['(NA)'], '0')
new_gp['2006'] = new_gp['2006'].replace(['(NA)'], '0')
new_gp['2005'] = new_gp['2005'].replace(['(NA)'], '0')

In [54]:
new_gp['2005']=pd.to_numeric(new_gp['2005'], errors='coerce')
new_gp['2006']=pd.to_numeric(new_gp['2006'], errors='coerce')
new_gp['2007']=pd.to_numeric(new_gp['2007'], errors='coerce')
new_gp['2008']=pd.to_numeric(new_gp['2008'], errors='coerce')
new_gp['2009']=pd.to_numeric(new_gp['2009'], errors='coerce')
new_gp['2010']=pd.to_numeric(new_gp['2010'], errors='coerce')
new_gp['2011']=pd.to_numeric(new_gp['2011'], errors='coerce')
new_gp['2012']=pd.to_numeric(new_gp['2012'], errors='coerce')
new_gp['2013']=pd.to_numeric(new_gp['2013'], errors='coerce')
new_gp['2014']=pd.to_numeric(new_gp['2014'], errors='coerce')
new_gp['2015']=pd.to_numeric(new_gp['2015'], errors='coerce')
new_gp['2016']=pd.to_numeric(new_gp['2016'], errors='coerce')
new_gp['2017']=pd.to_numeric(new_gp['2017'], errors='coerce')

In [55]:
new_gp = new_gp.where((pd.notnull(new_gp)), 0)
new_gp.isnull().sum()

State    0
2005     0
2006     0
2007     0
2008     0
2009     0
2010     0
2011     0
2012     0
2013     0
2014     0
2015     0
2016     0
2017     0
dtype: int64

In [56]:
# Check to see if data is complete.
new_gp.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 5400 entries, 0 to 5399
Data columns (total 14 columns):
State    5400 non-null object
2005     5400 non-null float64
2006     5400 non-null float64
2007     5400 non-null float64
2008     5400 non-null float64
2009     5400 non-null float64
2010     5400 non-null float64
2011     5400 non-null float64
2012     5400 non-null float64
2013     5400 non-null float64
2014     5400 non-null float64
2015     5400 non-null float64
2016     5400 non-null float64
2017     5400 non-null float64
dtypes: float64(13), object(1)
memory usage: 632.8+ KB


In [57]:
# Check format of dataframe
new_gp.head()

State        2005        2006        2007        2008        2009  \
0  United States  13015161.0  13773226.0  14391149.0  14626598.0  14320114.0   
1  United States  11366297.0  12045446.0  12572387.0  12716179.0  12352979.0   
2  United States    128571.0    128345.0    141999.0    154525.0    137655.0   
3  United States    104514.0     99352.0    113533.0    126345.0    109800.0   
4  United States     24057.0     28993.0     28466.0     28180.0     27855.0   

         2010        2011        2012        2013        2014        2015  \
0  14859772.0  15406002.0  16041243.0  16576738.0  17312308.0  18007206.0   
1  12826507.0  13348439.0  13957545.0  14468465.0  15149621.0  15776274.0   
2    160217.0    197241.0    185800.0    221821.0    204404.0    184791.0   
3    129725.0    166249.0    151489.0    186960.0    167709.0    145476.0   
4     30492.0     30992.0     34311.0     34861.0     36695.0     39315.0   

         2016        2017  
0  18509998.0  19263350.0  
1  16224645.0  16925936.0  
2    177580.0    173445.0  
3    136672.0         0.0  
4     40907.0         0.0

## Sum State and annual GDP and create new dataframe

In [58]:
tot_gp = new_gp.groupby('State').sum()
tot_gp.head()

2005       2006        2007        2008        2009  \
State                                                                  
Alabama      808437.0   847602.0    872058.0    886407.0    858404.0   
Alaska       207516.0   233876.0    259668.0    296315.0    264511.0   
Arizona     1182301.0  1289785.0   1362237.0   1335204.0   1262204.0   
Arkansas     472125.0   495655.0    508383.0    521646.0    509153.0   
California  9276216.0  9876594.0  10283245.0  10452769.0  10060201.0   

                  2010        2011        2012        2013        2014  \
State                                                                    
Alabama       894232.0    928116.0    958764.0    985504.0   1004334.0   
Alaska        285414.0    311510.0    323021.0    317203.0    306051.0   
Arizona      1281732.0   1329067.0   1387248.0   1417447.0   1475845.0   
Arkansas      535879.0    561132.0    569753.0    599610.0    618974.0   
California  10343486.0  10729064.0  11263961.0  11747101.0  12475325.0   

                  2015        2016        2017  
State                                           
Alabama      1034454.0   1056452.0    931408.0  
Alaska        270512.0    259129.0    227946.0  
Arizona      1542968.0   1602386.0   1444232.0  
Arkansas      623966.0    635156.0    565369.0  
California  13263037.0  13839799.0  12338557.0

## Turn year columns to rows within state to match with Data Breach file

In [59]:
ntot_gp = pd.melt(tot_gp.reset_index(), id_vars=['State'], var_name=['Breach_Year'])
ntot_gp.head()

State Breach_Year      value
0     Alabama        2005   808437.0
1      Alaska        2005   207516.0
2     Arizona        2005  1182301.0
3    Arkansas        2005   472125.0
4  California        2005  9276216.0

## sort new dataframe into state, year which is same order as data breach file

In [60]:
ntot_gp = ntot_gp.sort_values(["State","Breach_Year"], ascending=True)  
ntot_gp.head()

State Breach_Year     value
0    Alabama        2005  808437.0
60   Alabama        2006  847602.0
120  Alabama        2007  872058.0
180  Alabama        2008  886407.0
240  Alabama        2009  858404.0

In [61]:
ntot_gp['State_CAT'] = ntot_gp['State'].map({'Wyoming' : '52', 'Wisconsin' : '51', 'West Virginia' : '50', 'Washington' : '49', 'Virginia' : '48', 'Vermont' : '47', 'Utah' : '46', 'Texas' : '45', 'Tennessee' : '44', 'South Dakota' : '43', 'South Carolina' : '42', 'Rhode Island' : '41', 'Puerto Rico' : '40', 'Pennsylvania' : '39', 'Oregon' : '38', 'Oklahoma' : '37', 'Ohio' : '36', 'North Dakota' : '35', 'North Carolina' : '34', 'New York' : '33', 'New Mexico' : '32', 'New Jersey' : '31', 'New Hampshire' : '30', 'Nevada' : '29', 'Nebraska' : '28', 'Montana' : '27', 'Missouri' : '26', 'Mississippi' : '25', 'Minnesota' : '24', 'Michigan' : '23', 'Massachusetts' : '22', 'Maryland' : '21', 'Maine' : '20', 'Louisiana' : '19', 'Kentucky' : '18', 'Kansas' : '17', 'Iowa' : '16', 'Indiana' : '15', 'Illinois' : '14', 'Idaho' : '13', 'Hawaii' : '12', 'Georgia' : '11', 'Florida' : '10', 'District of Columbia' : '9', 'Delaware': '8', 'United States' : '0', 'Alabama': '1', 'Alaska': '2', 'Arizona': '3', 'Arkansas': '4', 'California': '5', 'Colorado': '6', 'Connecticut': '7'})

In [62]:
ntot_gp['Year_CAT'] = ntot_gp['Breach_Year'].map({'2005' : '1', '2006' : '2', '2007' : '3', '2008' : '4', '2009' : '5', '2010' : '6', '2011' : '7', '2012' : '8', '2013' : '9', '2014' : '10', '2015' : '11', '2016' : '12', '2017' : '13', '2018' : '14'})

In [63]:
ntot_gp

State Breach_Year      value State_CAT Year_CAT
0          Alabama        2005   808437.0         1        1
60         Alabama        2006   847602.0         1        2
120        Alabama        2007   872058.0         1        3
180        Alabama        2008   886407.0         1        4
240        Alabama        2009   858404.0         1        5
300        Alabama        2010   894232.0         1        6
360        Alabama        2011   928116.0         1        7
420        Alabama        2012   958764.0         1        8
480        Alabama        2013   985504.0         1        9
540        Alabama        2014  1004334.0         1       10
600        Alabama        2015  1034454.0         1       11
660        Alabama        2016  1056452.0         1       12
720        Alabama        2017   931408.0         1       13
1           Alaska        2005   207516.0         2        1
61          Alaska        2006   233876.0         2        2
121         Alaska        2007   259668.0         2        3
181         Alaska        2008   296315.0         2        4
241         Alaska        2009   264511.0         2        5
301         Alaska        2010   285414.0         2        6
361         Alaska        2011   311510.0         2        7
421         Alaska        2012   323021.0         2        8
481         Alaska        2013   317203.0         2        9
541         Alaska        2014   306051.0         2       10
601         Alaska        2015   270512.0         2       11
661         Alaska        2016   259129.0         2       12
721         Alaska        2017   227946.0         2       13
2          Arizona        2005  1182301.0         3        1
62         Arizona        2006  1289785.0         3        2
122        Arizona        2007  1362237.0         3        3
182        Arizona        2008  1335204.0         3        4
..             ...         ...        ...       ...      ...
597  West Virginia        2014   384193.0        50       10
657  West Virginia        2015   376450.0        50       11
717  West Virginia        2016   375985.0        50       12
777  West Virginia        2017   340775.0        50       13
58       Wisconsin        2005  1204120.0        51        1
118      Wisconsin        2006  1255216.0        51        2
178      Wisconsin        2007  1296452.0        51        3
238      Wisconsin        2008  1297291.0        51        4
298      Wisconsin        2009  1297886.0        51        5
358      Wisconsin        2010  1342817.0        51        6
418      Wisconsin        2011  1397644.0        51        7
478      Wisconsin        2012  1451477.0        51        8
538      Wisconsin        2013  1502690.0        51        9
598      Wisconsin        2014  1553906.0        51       10
658      Wisconsin        2015  1623285.0        51       11
718      Wisconsin        2016  1678992.0        51       12
778      Wisconsin        2017  1484478.0        51       13
59         Wyoming        2005   145814.0        52        1
119        Wyoming        2006   175109.0        52        2
179        Wyoming        2007   194560.0        52        3
239        Wyoming        2008   231995.0        52        4
299        Wyoming        2009   197468.0        52        5
359        Wyoming        2010   208703.0        52        6
419        Wyoming        2011   222871.0        52        7
479        Wyoming        2012   213460.0        52        8
539        Wyoming        2013   216985.0        52        9
599        Wyoming        2014   219279.0        52       10
659        Wyoming        2015   206896.0        52       11
719        Wyoming        2016   197807.0        52       12
779        Wyoming        2017   179461.0        52       13

[780 rows x 5 columns]

In [64]:
ntot_gp.groupby('State').sum()

value
State                             
Alabama               1.206617e+07
Alaska                3.562672e+06
Arizona               1.791266e+07
Arkansas              7.216801e+06
California            1.459494e+08
Colorado              1.845718e+07
Connecticut           1.647092e+07
Delaware              4.293065e+06
District of Columbia  6.379661e+06
Far West              2.021435e+08
Florida               5.345010e+07
Georgia               3.003763e+07
Great Lakes           1.525032e+08
Hawaii                4.622855e+06
Idaho                 3.930937e+06
Illinois              4.780682e+07
Indiana               2.057460e+07
Iowa                  1.057116e+07
Kansas                9.079061e+06
Kentucky              1.155588e+07
Louisiana             1.565007e+07
Maine                 3.546288e+06
Maryland              2.101208e+07
Massachusetts         2.913712e+07
Michigan              2.877740e+07
Mideast               1.953845e+08
Minnesota             1.978430e+07
Mississippi           6.481004e+06
Missouri              1.783357e+07
Montana               2.684374e+06
Nebraska              6.661694e+06
Nevada                8.944035e+06
New England           5.896707e+07
New Hampshire         4.545022e+06
New Jersey            3.490838e+07
New Mexico            5.628473e+06
New York              8.623194e+07
North Carolina        2.979065e+07
North Dakota          2.893647e+06
Ohio                  3.695809e+07
Oklahoma              1.104100e+07
Oregon                1.321760e+07
Pennsylvania          4.258897e+07
Plains                6.961742e+07
Rhode Island          3.427123e+06
Rocky Mountain        3.623227e+07
South Carolina        1.166107e+07
South Dakota          2.793978e+06
Southeast             2.294078e+08
Southwest             1.279599e+08
Tennessee             1.872020e+07
Texas                 9.337780e+07
United States         1.072249e+09
Utah                  8.549364e+06
Vermont               1.844044e+06
Virginia              2.831108e+07
Washington            2.584698e+07
West Virginia         4.467215e+06
Wisconsin             1.838625e+07
Wyoming               2.610408e+06

In [65]:
ntot_gp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 780 entries, 0 to 779
Data columns (total 5 columns):
State          780 non-null object
Breach_Year    780 non-null object
value          780 non-null float64
State_CAT      676 non-null object
Year_CAT       780 non-null object
dtypes: float64(1), object(4)
memory usage: 36.6+ KB


## Will be combining the breach file and the gdp file by state by year

- Join stot_gp and df_grp by State and Breach_Year.  
- Normalize counts by state/year.  
- Create a factor and adjust occurrence. 
- Join to Databreach file.

In [66]:
# replace null value with 'United States' 
ntot_gp = ntot_gp.fillna({'State_CAT': '99'})

In [67]:
ntot_gp = ntot_gp.where((pd.notnull(ntot_gp)), '0')
ntot_gp.isnull().sum()

State          0
Breach_Year    0
value          0
State_CAT      0
Year_CAT       0
dtype: int64

In [68]:
#Convert Category to integer
ntot_gp['State_CAT'] = ntot_gp['State_CAT'].astype(np.int64)

In [69]:
#Convert Category to integer
ntot_gp['Year_CAT'] = ntot_gp['Year_CAT'].astype(np.int64)

In [70]:
stot_gp = ntot_gp.sort_values(["State_CAT","Year_CAT"], ascending=True)  
stot_gp

State Breach_Year       value  State_CAT  Year_CAT
52    United States        2005  68421336.0          0         1
112   United States        2006  72466970.0          0         2
172   United States        2007  75725457.0          0         3
232   United States        2008  76947951.0          0         4
292   United States        2009  75174436.0          0         5
352   United States        2010  78138882.0          0         6
412   United States        2011  81236392.0          0         7
472   United States        2012  84735564.0          0         8
532   United States        2013  87628248.0          0         9
592   United States        2014  91612470.0          0        10
652   United States        2015  95216115.0          0        11
712   United States        2016  97832555.0          0        12
772   United States        2017  87113022.0          0        13
0           Alabama        2005    808437.0          1         1
60          Alabama        2006    847602.0          1         2
120         Alabama        2007    872058.0          1         3
180         Alabama        2008    886407.0          1         4
240         Alabama        2009    858404.0          1         5
300         Alabama        2010    894232.0          1         6
360         Alabama        2011    928116.0          1         7
420         Alabama        2012    958764.0          1         8
480         Alabama        2013    985504.0          1         9
540         Alabama        2014   1004334.0          1        10
600         Alabama        2015   1034454.0          1        11
660         Alabama        2016   1056452.0          1        12
720         Alabama        2017    931408.0          1        13
1            Alaska        2005    207516.0          2         1
61           Alaska        2006    233876.0          2         2
121          Alaska        2007    259668.0          2         3
181          Alaska        2008    296315.0          2         4
..              ...         ...         ...        ...       ...
565         Mideast        2014  16626691.0         99        10
572     New England        2014   4870574.0         99        10
583          Plains        2014   6053356.0         99        10
585  Rocky Mountain        2014   3138240.0         99        10
588       Southeast        2014  19167479.0         99        10
589       Southwest        2014  11659947.0         99        10
609        Far West        2015  18218444.0         99        11
612     Great Lakes        2015  13395871.0         99        11
625         Mideast        2015  17350913.0         99        11
632     New England        2015   5132084.0         99        11
643          Plains        2015   6222309.0         99        11
645  Rocky Mountain        2015   3234238.0         99        11
648       Southeast        2015  20072447.0         99        11
649       Southwest        2015  11589795.0         99        11
669        Far West        2016  19008243.0         99        12
672     Great Lakes        2016  13776350.0         99        12
685         Mideast        2016  17841540.0         99        12
692     New England        2016   5274937.0         99        12
703          Plains        2016   6369345.0         99        12
705  Rocky Mountain        2016   3320385.0         99        12
708       Southeast        2016  20739400.0         99        12
709       Southwest        2016  11502348.0         99        12
729        Far West        2017  16974855.0         99        13
732     Great Lakes        2017  12243056.0         99        13
745         Mideast        2017  15693702.0         99        13
752     New England        2017   4667193.0         99        13
763          Plains        2017   5621988.0         99        13
765  Rocky Mountain        2017   3000073.0         99        13
768       Southeast        2017  18428522.0         99        13
769       Southwest        2017  10449993.0      

In [71]:
stot_gp = stot_gp[stot_gp.State_CAT != 99]


In [72]:
stot_gp = stot_gp.drop(["State","Breach_Year"], axis=1)
stot_gp.head()

value  State_CAT  Year_CAT
52   68421336.0          0         1
112  72466970.0          0         2
172  75725457.0          0         3
232  76947951.0          0         4
292  75174436.0          0         5

In [73]:
df = df.sort_values(["State_CAT","Year_CAT"], ascending=True)  
df

Breach_Year                                            Company  \
8136        2006                          Deloitte & Touche, McAfee   
8134        2007                      TennCare, Americhoice Inc.      
8139        2007                         Mercer Health and Benefits   
8140        2007     Fidelity Investments, Dairy Farmers of America   
8141        2007                                         Dai Nippon   
8142        2007                       Verus Inc., Concord Hospital   
8138        2008  SunGard, Connecticut State University System, ...   
8135        2009                                     U.S. Consulate   
8137        2009                        Eastern Illinois University   
8176        2010                                             Mercer   
8177        2010  Puerto Rico Department of Health - Triple S Ma...   
8178        2010                               Triple-S Salud, Inc.   
8179        2010                             Hospital Auxilio Mutuo   
8143        2011  Rolling Stone, Radar, Corrupted Justice, Netti...   
8144        2011  SettlementOne Credit Corporation, Sackett Nati...   
8145        2011                           Catholic Social Services   
8146        2011          Park Avenue Obstetrics and Gynecology, PC   
8147        2011                                       Regions Bank   
8148        2011                    Ohio Auditor of State Dave Yost   
8149        2011                             HarborOne Credit Union   
8150        2011                                     DSLReports.com   
8151        2011  International Association of Chiefs of Police ...   
8152        2011                                 Pan American Games   
8153        2011                       Atlanta Perinatal Associates   
8154        2011          Stone Oak Urgent Care and Family Practice   
8155        2011     State of Tennessee Sponsored Group Health Plan   
8159        2011   Fairview and North Memorial Hospitals, Accretive   
8170        2011  TRICARE Management Activity (formerly Civilian...   
8180        2011               Departamento de Salud de Puerto Rico   
8181        2011                                        MAPFRE Life   
...          ...                                                ...   
8087        2017                         Metropolitan Urology Group   
8105        2017                             SSM Dean Medical Group   
8106        2017                                          UW Health   
8107        2017                                          UW Health   
8108        2017                                 Community Link Inc   
8109        2017                                     Network Health   
8110        2017                  University of Wisconsin - Madison   
8111        2017                      Briggs & Stratton Corporation   
8112        2017                       Medical College of Wisconsin   
8113        2017                                     Kimberly-Clark   
8114        2017             The Medical College of Wisconsin, Inc.   
8115        2018                                            QuadMed   
8116        2018                                      Quad/Med, LLC   
8117        2018                                  Doral Corporation   
8118        2018            Wisconsin Department of Health Services   
8119        2018                    Scenic Bluffs Health Center Inc   
8120        2018                               RISE Wisconsin, Inc.   
8121        2009                                Rocky Mountain Bank   
8122        2010                       Wyoming Department of Health   
8123        2010         Wyoming Department of Health Kid Care CHIP   
8124        2011     Community Action Partnership of Natrona County   
8127        2011     Community Action partnership of Natrona County   
8128        2013                              Hansen and Associates   
8129        2013                       Wyoming Department of Health   
8130        2014             

In [74]:
df = df.drop(["Breach_Occurrence"], axis=1)
df.head()

Breach_Year                                         Company  \
8136        2006                       Deloitte & Touche, McAfee   
8134        2007                   TennCare, Americhoice Inc.      
8139        2007                      Mercer Health and Benefits   
8140        2007  Fidelity Investments, Dairy Farmers of America   
8141        2007                                      Dai Nippon   

              State Breach_Type Organization_Type  Total_Recs  \
8136  United States        PORT               BSO        9290   
8134  United States        PORT               MED       67000   
8139  United States        PORT               BSF       10500   
8140  United States        PORT               BSF          69   
8141  United States        INSD               BSO           0   

                                            Description   Latitude  Longitude  \
8136  An external auditor lost a CD with names, Soci...  37.090240 -95.712891   
8134  There are 67,000 TennCare \n            enroll...  35.960638 -83.920739   
8139     A laptop computer was stolen from a vehicle.    37.090240 -95.712891   
8140  A Fidelity laptop used by a former Fidelity em...  37.090240 -95.712891   
8141  A former contract worker of a Japanese commerc...  37.090240 -95.712891   

      Breach_Type_CAT  Organization_Type_CAT  State_CAT  Year_CAT  
8136                5                      2          0         2  
8134                5                      6          0         3  
8139                5                      1          0         3  
8140                5                      1          0         3  
8141                3                      2          0         3

In [75]:
#merge the GDP with the data breach file
df_new = pd.merge(df, stot_gp, on=["State_CAT", "Year_CAT"], how="left")
df_new  

Breach_Year                                            Company  \
0           2006                          Deloitte & Touche, McAfee   
1           2007                      TennCare, Americhoice Inc.      
2           2007                         Mercer Health and Benefits   
3           2007     Fidelity Investments, Dairy Farmers of America   
4           2007                                         Dai Nippon   
5           2007                       Verus Inc., Concord Hospital   
6           2008  SunGard, Connecticut State University System, ...   
7           2009                                     U.S. Consulate   
8           2009                        Eastern Illinois University   
9           2010                                             Mercer   
10          2010  Puerto Rico Department of Health - Triple S Ma...   
11          2010                               Triple-S Salud, Inc.   
12          2010                             Hospital Auxilio Mutuo   
13          2011  Rolling Stone, Radar, Corrupted Justice, Netti...   
14          2011  SettlementOne Credit Corporation, Sackett Nati...   
15          2011                           Catholic Social Services   
16          2011          Park Avenue Obstetrics and Gynecology, PC   
17          2011                                       Regions Bank   
18          2011                    Ohio Auditor of State Dave Yost   
19          2011                             HarborOne Credit Union   
20          2011                                     DSLReports.com   
21          2011  International Association of Chiefs of Police ...   
22          2011                                 Pan American Games   
23          2011                       Atlanta Perinatal Associates   
24          2011          Stone Oak Urgent Care and Family Practice   
25          2011     State of Tennessee Sponsored Group Health Plan   
26          2011   Fairview and North Memorial Hospitals, Accretive   
27          2011  TRICARE Management Activity (formerly Civilian...   
28          2011               Departamento de Salud de Puerto Rico   
29          2011                                        MAPFRE Life   
...          ...                                                ...   
8147        2017                         Metropolitan Urology Group   
8148        2017                             SSM Dean Medical Group   
8149        2017                                          UW Health   
8150        2017                                          UW Health   
8151        2017                                 Community Link Inc   
8152        2017                                     Network Health   
8153        2017                  University of Wisconsin - Madison   
8154        2017                      Briggs & Stratton Corporation   
8155        2017                       Medical College of Wisconsin   
8156        2017                                     Kimberly-Clark   
8157        2017             The Medical College of Wisconsin, Inc.   
8158        2018                                            QuadMed   
8159        2018                                      Quad/Med, LLC   
8160        2018                                  Doral Corporation   
8161        2018            Wisconsin Department of Health Services   
8162        2018                    Scenic Bluffs Health Center Inc   
8163        2018                               RISE Wisconsin, Inc.   
8164        2009                                Rocky Mountain Bank   
8165        2010                       Wyoming Department of Health   
8166        2010         Wyoming Department of Health Kid Care CHIP   
8167        2011     Community Action Partnership of Natrona County   
8168        2011     Community Action partnership of Natrona County   
8169        2013                              Hansen and Associates   
8170        2013                       Wyoming Department of Health   
8171        2014             

In [76]:
#fill df with zeros where null
df_new = df_new.where((pd.notnull(df_new)), '0')
df_new.isnull().sum()

Breach_Year              0
Company                  0
State                    0
Breach_Type              0
Organization_Type        0
Total_Recs               0
Description              0
Latitude                 0
Longitude                0
Breach_Type_CAT          0
Organization_Type_CAT    0
State_CAT                0
Year_CAT                 0
value                    0
dtype: int64

In [77]:
#Convert value to integer
df_new['value'] = df_new['value'].astype(np.int64)

In [78]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8177 entries, 0 to 8176
Data columns (total 14 columns):
Breach_Year              8177 non-null object
Company                  8177 non-null object
State                    8177 non-null object
Breach_Type              8177 non-null object
Organization_Type        8177 non-null object
Total_Recs               8177 non-null int64
Description              8177 non-null object
Latitude                 8177 non-null object
Longitude                8177 non-null object
Breach_Type_CAT          8177 non-null int64
Organization_Type_CAT    8177 non-null int64
State_CAT                8177 non-null int64
Year_CAT                 8177 non-null int64
value                    8177 non-null int64
dtypes: int64(6), object(8)
memory usage: 958.2+ KB


In [79]:
df_new.head()

Breach_Year                                         Company          State  \
0        2006                       Deloitte & Touche, McAfee  United States   
1        2007                   TennCare, Americhoice Inc.     United States   
2        2007                      Mercer Health and Benefits  United States   
3        2007  Fidelity Investments, Dairy Farmers of America  United States   
4        2007                                      Dai Nippon  United States   

  Breach_Type Organization_Type  Total_Recs  \
0        PORT               BSO        9290   
1        PORT               MED       67000   
2        PORT               BSF       10500   
3        PORT               BSF          69   
4        INSD               BSO           0   

                                         Description Latitude Longitude  \
0  An external auditor lost a CD with names, Soci...  37.0902  -95.7129   
1  There are 67,000 TennCare \n            enroll...  35.9606  -83.9207   
2     A laptop computer was stolen from a vehicle.    37.0902  -95.7129   
3  A Fidelity laptop used by a former Fidelity em...  37.0902  -95.7129   
4  A former contract worker of a Japanese commerc...  37.0902  -95.7129   

   Breach_Type_CAT  Organization_Type_CAT  State_CAT  Year_CAT     value  
0                5                      2          0         2  72466970  
1                5                      6          0         3  75725457  
2                5                      1          0         3  75725457  
3                5                      1          0         3  75725457  
4                3                      2          0         3  75725457

In [80]:
# rename columns
df_new = df_new.rename(columns={'value': 'GDP'})
df_new.head()

Breach_Year                                         Company          State  \
0        2006                       Deloitte & Touche, McAfee  United States   
1        2007                   TennCare, Americhoice Inc.     United States   
2        2007                      Mercer Health and Benefits  United States   
3        2007  Fidelity Investments, Dairy Farmers of America  United States   
4        2007                                      Dai Nippon  United States   

  Breach_Type Organization_Type  Total_Recs  \
0        PORT               BSO        9290   
1        PORT               MED       67000   
2        PORT               BSF       10500   
3        PORT               BSF          69   
4        INSD               BSO           0   

                                         Description Latitude Longitude  \
0  An external auditor lost a CD with names, Soci...  37.0902  -95.7129   
1  There are 67,000 TennCare \n            enroll...  35.9606  -83.9207   
2     A laptop computer was stolen from a vehicle.    37.0902  -95.7129   
3  A Fidelity laptop used by a former Fidelity em...  37.0902  -95.7129   
4  A former contract worker of a Japanese commerc...  37.0902  -95.7129   

   Breach_Type_CAT  Organization_Type_CAT  State_CAT  Year_CAT       GDP  
0                5                      2          0         2  72466970  
1                5                      6          0         3  75725457  
2                5                      1          0         3  75725457  
3                5                      1          0         3  75725457  
4                3                      2          0         3  75725457

## Write Clean and Transformed File

In [81]:
df_new.to_csv("databreach_cleaned.csv",index=False)